File: ruby/grade-school/grade_school.rb
class School
  attr_accessor :all_students
  def initialize
    @all_students = []
  end

  def add(name, grade)
    @all_students << Student.new(name, grade)
  end

  def students(grade)
    @all_students.select { |e| e.grade == grade }.map { |e| e.name }.sort rescue []
  end

  def grades
    @grades ||= @all_students.map { |e| e.grade }.uniq.sort
  end

  def students_by_grade
    grades.map { |e| {grade: e, students: students(e)} }
  end

  class Student
    attr_reader :name, :grade
    def initialize(name, grade)
      @name = name
      @grade = grade
    end

  end

end

module BookKeeping
  VERSION = 3
end

File: python/grade-school/grade_school.py
class School(object):
    def __init__(self):
        self.school = {}

    def add_student(self, name, grade):
        if grade in self.school.keys():
            self.school[grade].append(name)
        else:
            self.school[grade] = [name]
        return self.school

    def roster(self):
        return [st for grade, students in sorted(self.school.items()) for st in sorted(students)]


    def grade(self, grade_number):
        if grade_number in self.school:
            return sorted(self.school[grade_number])
        else:
            return []
